c++/src/kj/compat/http.c++

```diff
@@ -3489,6 +3489,358 @@ WebSocketPipe newWebSocketPipe() {
+      } else {
+        // Invalid ABNF, expected 1*DIGIT.
+        return nullptr;
+      }
+    }
+  }
+
+  KJ_IF_MAYBE(clientBits, config.clientMaxWindowBits) {
+    if (clientBits->size() == 0) {
+      if (!isAgreement) {
+        // `client_max_window_bits` does not need to have a value in an offer, let's set it to 15
+        // to get the best level of compression.
+        result.inboundMaxWindowBits = 15;
+      } else {
+        // `client_max_window_bits` must have a value in a Response.
+        return nullptr;
+      }
+    } else {
+      KJ_IF_MAYBE(bits, extractBitsFromValue(kj::mv(*clientBits))) {
+        if (*bits < 8 || 15 < *bits) {
+          // Out of range -- invalid.
+          return nullptr;
+        }
+        if (isAgreement) {
+          result.outboundMaxWindowBits = *bits;
+        } else {
+          result.inboundMaxWindowBits = *bits;
+        }
+      } else {
+        // Invalid ABNF, expected 1*DIGIT.
+        return nullptr;
+      }
+    }
+  }
+
+  if (isAgreement) {
+    result.outboundNoContextTakeover = config.clientNoContextTakeover;
+    result.inboundNoContextTakeover = config.serverNoContextTakeover;
+  } else {
+    result.inboundNoContextTakeover = config.clientNoContextTakeover;
+    result.outboundNoContextTakeover = config.serverNoContextTakeover;
+  }
+  return kj::mv(result);
+}
+
+inline bool isValidExtensionName(StringPtr extension, StringPtr permitted) {
+  // Ensures the extension name is not followed by invalid characters.
+  // Assumes that `extension` starts with `permitted` (caller's responsibility).
+  KJ_REQUIRE(permitted.size() <= extension.size());
+
+  ArrayPtr<const char> tail = extension.slice(permitted.size(), extension.size());
+  stripLeadingSpace(tail);
+
+  return tail.size() == 0;
+}
+
+inline kj::Maybe<CompressionParameters> tryExtractParameters(kj::Vector<String>& configuration,
+    bool isAgreement) {
+  // If the `configuration` is structured correctly and has no invalid parameters/values, we will
+  // return a populated `CompressionParameters` struct.
+  if (configuration.size() == 1) {
+    // Plain `permessage-deflate`.
+    return CompressionParameters{};
+  }
+  auto params = configuration.slice(1, configuration.size());
+  auto keyMaybeValuePairs = toKeysAndVals(params);
+  // Parse parameter strings into parameter[=value] pairs.
+  auto maybeUnverified = populateUnverifiedConfig(keyMaybeValuePairs);
+  KJ_IF_MAYBE(unverified, maybeUnverified) {
+    // Parsing succeeded, i.e. the parameter (`key`) names are valid and we don't have
+    // values for `x_no_context_takeover` parameters (the configuration is structured correctly).
+    // All that's left is to check the `x_max_window_bits` values (if any are present).
+    KJ_IF_MAYBE(validConfig, validateCompressionConfig(kj::mv(*unverified), isAgreement)) {
+      return kj::mv(*validConfig);
+    }
+  }
+  return nullptr;
+}
+
+kj::Vector<CompressionParameters> findValidExtensionOffers(StringPtr offers) {
```

Comment:
Should add this to `http.h` so we can test it.

Reply:
https://github.com/capnproto/capnproto/pull/1480/commits/74580a60b546c599bf2aab71e2f86b52c0bbc8e5